code/network/lib/Loot/Network/ZMQ/Client.hs

```diff
@@ -191,37 +183,84 @@ data InternalRequest
     = IRUpdatePeers ZTUpdatePeersReq
     | IRRegister ClientId (Set MsgType) (Set Subscription) ZTClientEnv
     | IRReconnect (Set ZTNodeId)
+    deriving (Generic)
 
 instance T.Show InternalRequest where
     show (IRUpdatePeers r)      = "IRUpdatePeers: " <> show r
     show (IRRegister cId _ _ _) = "IRRegister " <> show cId
     show (IRReconnect nids)     = "IRReconnect" <> show nids
 
--- Get hashmap keys as a hashset.
-hmKeys :: HashMap k a -> HashSet k
-hmKeys = HSet.fromMap . HMap.map (const ())
+type CliRequestQueue = InternalQueue InternalRequest
+
+----------------------------------------------------------------------------
+-- Polling data
+----------------------------------------------------------------------------
 
--- | Wrapper over client request queue. CLient broker listens to this.
-newtype CliRequestQueue = CliRequestQueue
-    { unCliRequestQueue :: TQueue InternalRequest
+data CliPollData = CliPollData
+    { cpdPolls      :: ![Z.Poll Z.Socket IO]
+    , cpdN          :: !Int
```

Comment:
can't `cpdN` be computed as `length cpdPolls`?

Reply:
lool, speed, it's because you join all client workers to one thread, soo speedy

Reply:
sure there is place where you can compute length one time and reuse it, before `action` funtion

Reply:
>you join all client workers to one thread

Computing `cpdPolls` requires an STM action each poll, so yes, it's a noticeable overhead I believe.

>sure there is place where you can compute length one time and reuse it, before action funtion

Our peers list is not constant, it evolves over time.

Reply:
I believe that one iteration of 
`forM_ (events `zip` [0..]) $ \(e,i) ->`
takes more time than performing `length cpdPolls` after reading it from `TVar`, however, I don't want to dispute about it. let's it be in this way

Reply:
>I believe that one iteration of
forM_ (events zip [0..]) $ \(e,i) ->
takes more time than performing length cpdPolls

do you realise it?